commit 8c86e223bdf9be9bf0d25d15d8560656865e51a5 (refs/changes/92/392/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-08-31T18:02:26-07:00 (3 years, 1 month ago)
    
    AGENT-1034 cn-agent on old PIs pre-IMG.js thinks everything is an image

diff --git a/lib/heartbeater.js b/lib/heartbeater.js
index c8b7a92..4fee73e 100644
--- a/lib/heartbeater.js
+++ b/lib/heartbeater.js
@@ -496,7 +496,7 @@ StatusReporter.prototype.gatherDiskUsage = function (vms, callback) {
                 imgadm.quickGetImage({
                     uuid: datasetUuid,
                     log: self.log
-                }, function (err) {
+                }, function (err, manifest) {
                     if (err && err.code === 'ImageNotInstalled') {
                         next();
                         return;
@@ -508,6 +508,24 @@ StatusReporter.prototype.gatherDiskUsage = function (vms, callback) {
                         return;
                     }
 
+                    /*
+                     * If we used "imgadm get" we might get a dummy result for
+                     * a random UUID-named dataset. We can't easily tell whether
+                     * it's really an image (in the Bad Old Days no metadata was
+                     * really required), so if it has nothing but a uuid in its
+                     * manifest, skip it.
+                     *
+                     * It's better to miss one or two images here (they'll still
+                     * get counted against provisionable space by DAPI, just as
+                     * system_used instead) than to double-count real VMs
+                     * (which could make this box un-provisionable).
+                     */
+                    var keys = Object.keys(manifest);
+                    if (keys.length === 1 && keys[0] === 'uuid') {
+                        next();
+                        return;
+                    }
+
                     // Tally bytes used
                     usage.installed_images_used_bytes
                         += toInt(datasets[dataset].used);
